Improve upvote / downvote comment handling

paulmolluzzo opened the pull request:
This is a WIP to fix #32. It solves 2 of the 3 items, but I don't know how to solve the last one perfectly. If we decide to omit it then maybe this PR is done.

- [x] Smarter emoji detection

As of the opening of this PR :100: is an acceptable emoji to upvote. To add more acceptable emoji (or text) it can just be added to those arrays. @SpaceK33z set it up that way to begin with.

- [x] Count one vote per user. In a PR, a user can first down vote, but after some changes up vote. Only the up vote should count in this scenario.

Now each upvote/downvote is only counted once per user. If the same user upvotes, then downvotes, then upvotes, it should effectively count as a single upvote. Voting 100 times in either direction counts as 1 vote.

Not done:

- ~~If the upvote / downvote is followed by some text, don't remove the comment but let it count.~~

edit: not doing this since there are too many circumstances to consider. A comment with just a vote and nothing else is a vote. Everything else should stay.

It's not even just following text that's what you'll find. Sometimes the emoji can be at the end of the text. It can also be in the middle some place. And if you look at [this issue](https://github.com/dear-github/dear-github/issues/113) you can see that there's even people complaining out all the +1's with no content, and those aren't votes, just arguing about voting. Not sure how to solve that one.

hkdobrev commented on extension/content.js:
```diff
@@ -10,7 +10,7 @@ const isIssue = () => /^\/[^/]+\/[^/]+\/issues\/\d+$/.test(location.pathname);
 const getUsername = () => $('meta[name="user-login"]').attr('content');
 const uselessContent = {
-	upvote: {text: ['+1\n'], emoji: [':+1:']},
+	upvote: {text: ['+1\n'], emoji: [':+1:', ':100:']},
```
You can add :ok_hand: as well.

hkdobrev commented:
A couple of points:

- When a comment begins with a mention, I'd say not to count that as upvote or downvote disregarding of the content or emoji. This is most probably a reply to something. Sometimes someone can express negative opinion on something and someone to support this negative opinion by mentioning him and give a :+1: to his comment.
- I second what @SpaceK33z mentioned in https://github.com/sindresorhus/refined-github/issues/32#issuecomment-188423846. GitHub said they will add some changes to issues which could affect these +/- 1 interactions in substantial way.

paulmolluzzo commented:
I added in the acceptance of :ok_hand: as an upvote and also fixed a bug I found. I also want to point out why this PR is better than just leaving the current implementation. A picture is worth 1000 words, so here are screenshots highlighting the differences. In all of these I'm the only user upvoting and downvoting a bunch of times. I think the PR is a marked improvement.

## No extension:
![no-extension](https://cloud.githubusercontent.com/assets/737065/13340472/828c6b14-dbfe-11e5-8345-20527dbed26b.png)

## Before the PR:
![before-this-pr](https://cloud.githubusercontent.com/assets/737065/13340523/ea664c00-dbfe-11e5-90de-3555328de367.png)

## After the plugin:
![after-this-pr](https://cloud.githubusercontent.com/assets/737065/13340525/ef1acf82-dbfe-11e5-8ff0-00ef7b9cd25e.png)

---

To @hkdobrev's feedback:

> When a comment begins with a mention, I'd say not to count that as upvote or downvote disregarding of the content or emoji. This is most probably a reply to something. Sometimes someone can express negative opinion on something and someone to support this negative opinion by mentioning him and give a :+1: to his comment.

This PR doesn't change the way @SpaceK33z decided to include/omit comments for votes. The logic in the `commentIsUseless` function is that if a comment has nothing but an upvote/downvote it counts, and I left that alone, except now I return `false` for all non-useless comments to help with some other logic.

I think @SpaceK33z made a good choice there. :+1: 

> I second what @SpaceK33z mentioned in #32 (comment). GitHub said they will add some changes to issues which could affect these +/- 1 interactions in substantial way.

As I understand it, that's sort of the point of this extension. If it breaks because GitHub is improved, so be it.

Relevant part in the README:

> My hope is that GitHub will notice and implement some of these much needed improvements. So if you like any of these features, please email GitHub support about adding it.

sindresorhus commented on extension/content.js:
```diff
@@ -67,24 +69,41 @@ function renderVoteCount(type, count) {
+
 		const isUp = commentIsUseless('upvote', el);
 		const isDown = commentIsUseless('downvote', el);
+		const commenter = el.closest('.js-comment-container').getElementsByClassName('author')[0].innerHTML;
```
Why not just

```js
const commenter = el.closest('.js-comment-container').find('author').get(0).innerHTML;
```

paulmolluzzo replied:
I don't see a big difference, but I changed it.

paulmolluzzo commented:
@sindresorhus I think this is complete actually. I don't want to get into comment parsing to leave text or consider the context of the content. That's super complex I think. What about using +1 in a math equation? Or when someone does +1000! Or a phrase like "if you actually make this work then :+1:"?

A comment with just a vote and nothing else is a vote. Everything else should stay. What do you think?

sindresorhus commented:
> A comment with just a vote and nothing else is a vote. Everything else should stay. What do you think?

:+1:

paulmolluzzo commented:
OK, updated title and first comment. This should be OK to do a review and merge if it's good.

sindresorhus commented:
@paulmolluzzo Good stuff :)
